pkg/client/unversioned/clientcmd/merged_client_builder.go

```diff
@@ -115,10 +115,10 @@ func (config *DeferredLoadingClientConfig) ClientConfig() (*restclient.Config, e
 		// TODO: this shouldn't be a global - the client config rules should be
 		//   handling this.
 		defaultConfig, defErr := DefaultClientConfig.ClientConfig()
-		if IsConfigurationInvalid(defErr) && !IsEmptyConfig(err) {
+		if defErr != nil && !IsConfigurationInvalid(defErr) {
 			return mergedConfig, nil
```

Comment:
Here is an example test that fails - https://github.com/kubernetes/kubernetes/blob/master/test/e2e/federated-service.go#L102

> The default kubectl settings don't mandate in cluster config unless localhost:8080 is defined to be the cluster

This is not the behavior in `merged_client_builder.go` now or even after this PR, because `DefaultClientConfig` is never set to "https://localhost:8080".

Everything else there makes sense.

> Looking at BuildFromKubeconfigGetter, it's using Load() which returns an empty data on an empty secret. Can you point me to the test case that is failing (since Gubernator seems to be down)? Do you have a valid secret at that point? 

Yeah. I think so. It is possible to not have a secret at all in which case KubeconfigGetterForSecret.KubeconfigGetter() returns an error. It will not be empty however.

> But it also looks like you are doing some sort of InClusterConfig detection up front (https://github.com/kubernetes/kubernetes/blob/master/federation/pkg/federation-controller/util/cluster_util.go#L105) which probably should be part of clientcmd in the future.

Sure. Noted. You probably know this already, but it is worth pointing out (for posterity) that the client which is obtained by calling `client.NewInCluster()` is different than the client for which we are fetching the kubeconfig here.